fdbserver/ClusterController.actor.cpp

```diff
@@ -3295,108 +3365,166 @@ void checkOutstandingStorageRequests(ClusterControllerData* self) {
-		     newDDWorker.interf.locality.processId() != self->masterProcessId) ||
-		    (ddFitness == bestFitnessForDD &&
-		     newRKWorker.interf.locality.processId() != newDDWorker.interf.locality.processId() && ratekeeperHealthy &&
-		     currentRKProcessId.present() && currentDDProcessId == currentRKProcessId &&
-		     (newRKWorker.interf.locality.processId() != self->masterProcessId &&
-		      newDDWorker.interf.locality.processId() != self->masterProcessId))) {
-			TraceEvent("CCHaltDD", self->id)
-			    .detail("DDID", db.distributor.get().id())
-			    .detail("Excluded", ddWorker.priorityInfo.isExcluded)
-			    .detail("Fitness", ddFitness)
-			    .detail("BestFitness", bestFitnessForDD)
-			    .detail("CurrentRateKeeperProcessId",
-			            currentRKProcessId.present() ? currentRKProcessId.get() : LiteralStringRef("None"))
-			    .detail("CurrentDDProcessId", currentDDProcessId)
-			    .detail("MasterProcessID", self->masterProcessId)
-			    .detail("NewRKWorkers", newRKWorker.interf.locality.processId())
-			    .detail("NewDDWorker", newDDWorker.interf.locality.processId());
-			ddWorker.haltDistributor = brokenPromiseToNever(
-			    db.distributor.get().haltDataDistributor.getReply(HaltDataDistributorRequest(self->id)));
+	auto rkSingleton = RatekeeperSingleton(db.ratekeeper);
+	auto ddSingleton = DataDistributorSingleton(db.distributor);
+
+	// Check if the singletons are healthy.
+	// side effect: try to rerecruit the singletons to more optimal processes
+	bool rkHealthy = isHealthySingleton<RatekeeperInterface>(
+	    self, newRKWorker, rkSingleton, bestFitnessForRK, self->recruitingRatekeeperID);
+
+	bool ddHealthy = isHealthySingleton<DataDistributorInterface>(
+	    self, newDDWorker, ddSingleton, bestFitnessForDD, self->recruitingDistributorID);
+
+	// if any of the singletons are unhealthy (rerecruited or not stable), then do not
+	// consider any further re-recruitments
+	if (!(rkHealthy && ddHealthy)) {
+		return;
+	}
+
+	// if we reach here, we know that the singletons are healthy so let's
+	// check if we can colocate the singletons in a more optimal way
+
+	// TODO: verify that we don't need to get the pid from the worker like we were doing before
+	Optional<Standalone<StringRef>> currRKProcessId = rkSingleton.interface.get().locality.processId();
+	Optional<Standalone<StringRef>> currDDProcessId = ddSingleton.interface.get().locality.processId();
+	Optional<Standalone<StringRef>> newRKProcessId = newRKWorker.interf.locality.processId();
+	Optional<Standalone<StringRef>> newDDProcessId = newRKWorker.interf.locality.processId();
+
+	auto currColocMap = getColocCounts({ currRKProcessId, currDDProcessId });
+	auto newColocMap = getColocCounts({ newRKProcessId, newDDProcessId });
+
+	// if the new coloc counts are not worse (i.e. each singleton's coloc count has not increased)
+	if (newColocMap[newRKProcessId] <= currColocMap[currRKProcessId] &&
```

Comment:
@sfc-gh-jslocum Here's the fix for the comparison between the old and new recruitment!